SI-6443 Widen dependent param types in uncurry

retronym opened the pull request:
Bridge building operates on unusual method signatures:
after uncurry, so parameter lists are collapsed; but before
erasure, so dependently typed parameters are still around.

Original:

```
def foo(a: T)(b: a.type, c: a.U): Unit
```

During computeBridges:

```
(a: T, b: a.type, c: a.U)Unit
```

This signature no longer appears to override the corresponding
one in a superclass, because the types of `b` and `c` are dependent
on method parameters.

The root of the problem is uncurry, which leaves the trees in
a poor state. This commit changes uncurry to remedy this.

An example illustrates it best:

```
// source
def foo(a: A)(b: a.type): b.type = b

// post uncurry before this patch.
// not well typed code!
def foo(a: A, b: a.type): b.type = {

// post uncurry after this patch
def foo(a: A, b: A): b.type = {
   val b$1 = b.asInstanceOf[a.type]
   b$1
}
```

This PR replaces #1696, and also closes SI-6135

Review by @adriaanm

retronym committed:
SI-6443 Widen dependent param types in uncurry

Bridge building operates on unusual method signatures:
after uncurry, so parameter lists are collapsed; but before
erasure, so dependently typed parameters are still around.

Original:

  def foo(a: T)(b: a.type, c: a.U): Unit

During computeBridges:

  (a: T, b: a.type, c: a.U)Unit

This signature no longer appears to override the corresponding
one in a superclass, because the types of `b` and `c` are dependent
on method parameters.

The root of the problem is uncurry, which leaves the trees in
a poor state. This commit changes uncurry to remedy this.

An example illustrates it best:

    // source
    def foo(a: A)(b: a.type): b.type = b

    // post uncurry before this patch.
    // not well typed code!
    def foo(a: A, b: a.type): a.type = {

    // post uncurry after this patch
    def foo(a: A, b: A): A = {
       val b$1 = b.asInstanceOf[a.type]
       b$1
    }

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2035/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2035/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1321/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1321/

adriaanm commented:
from the commit message `def foo(a: A)(b: A): a.type = {`
shouldn't this be `def foo(a: A, b: A): a.type = {` ?

adriaanm commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -746,15 +746,22 @@ abstract class UnCurry extends InfoTransform
             rhs = nonLocalReturnKeys get dd.symbol match {
-              case Some(k) => atPos(rhs0.pos)(nonLocalReturnTry(rhs0, k, dd.symbol))
-              case None    => rhs0
+              case Some(k) => atPos(rhs0.pos)(nonLocalReturnTry(newRhs, k, dd.symbol))
```
functionally equivalent, but less confusing: use `newRhs.pos` instead of `rhs0.pos`?

adriaanm commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -780,6 +787,110 @@ abstract class UnCurry extends InfoTransform
+      def erase(dd: DefDef): (List[List[ValDef]], Tree) = {
+        import dd.{ vparamss, rhs }
+        val vparamSyms = vparamss flatMap (_ map (_.symbol))
+        def eraseDependentType(declaredType: Type): Type = {
```
I think this should use ErasureMap:

Experimenting in :power mode using a slightly contrived example shows where the substitution-approach gives unexpected results (TypeBounds).

```
scala> class C { def x: T forSome {type T <: S; type S <: String} = ??? }
scala> val x = intp("C").info.decls.toList(1)
scala> x.info
=> T forSome { type T <: String; type S <: String }

scala> erasure.scalaErasure(x.info)
=> String

scala> val substed = x.info.resultType match {
     |   case ExistentialType(tparams, result) =>
     |     result.subst(tparams, tparams map (_.info))
     |   case tp => tp
     | }
  <: String

scala> substed.getClass
class scala.reflect.internal.Types$UniqueTypeBounds
```

adriaanm replied:
hmm, erasure probably taketh away too much -- maybe simply use widen on line 849?

retronym replied:
I don't think `widen` will cut it, we might have `def x(a: A)(b: a.B.C): (a.type, b.type)`.

But what I think we're after is `typer.packSymbols(params, resultTpe)`. With a slightly more intricate version of your example:

```
scala> class C { def x(a: String): T forSome {type T <: S; type S <: a.type} = ??? }
warning: there were 3 feature warnings; re-run with -feature for details
defined class C

scala> val x = intp("C").info.decls.toList(1)
x: $r.intp.global.Symbol = method x

scala> x.info
res6: $r.intp.global.Type = (a: String)T forSome { type T <: a.type; type S <: a.type }

scala> x.info.resultType
res7: $r.intp.global.Type = T forSome { type T <: a.type; type S <: a.type }

scala> x.info.resultType.widen
res8: $r.intp.global.Type = T forSome { type T <: a.type; type S <: a.type }

scala> typer.packSymbols(x.paramss.flatten, x.info.resultType)
res9: $r.intp.global.analyzer.global.Type = T forSome { val a: String; type T <: a.type; type S <: a.type }
```

adriaanm commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -780,6 +787,110 @@ abstract class UnCurry extends InfoTransform
+          vparamss.flatten.map { p =>
+            val declaredType = p.symbol.info
+            val widened = eraseDependentType(declaredType)
+            if (widened =:= declaredType) Unchanged(p)
```
using ErasureMap, I think you'll get exactly the same type (by reference) when nothing changes -- not 100% sure though

paulp replied:
Ha, you have somehow stumbled on exactly what ate up my yesterday. You totally will not get the same type by reference when nothing changes. You cannot rely on it in almost any case, for reasons I will tell you soon.

The fact that the compiler makes decisions based on the truth of (x.foo eq x) turns out to be at the root of some pretty long-running incorrectness.

adriaanm replied:
Good to know.

adriaanm replied:
I think we could just use `widen` instead of `eraseDependentType` here.

paulp replied:
See "the perils of reference equality" on scala-internals.

retronym replied:
Thanks for the suggestions, I'll try to expand the test case with the tougher variations and refine the way we cast off our dependencies.

adriaanm commented on src/compiler/scala/tools/nsc/transform/UnCurry.scala:
```diff
@@ -780,6 +787,110 @@ abstract class UnCurry extends InfoTransform
+
+        val allParams = widenings map (_.param)
+        val (widenedParams, tempVals) = widenings.collect {
+          case Widened(param, tempVal) => (param, tempVal)
```
fuse the map from line 877: `case Widened(param, tempVal) => (param.symbol, tempVal.symbol)`?

adriaanm replied:
ah shucks, tempVals is needed, so I guess saving the one map for widenedParams isn't worth the cost of asymmetry

retronym replied:
This code only runs for methods with dependent method types, so performance isn't so critical.

retronym committed:
SI-6443 Expand test coverage with varargs, by-name.

These were already working, due to a serendipitous
ordering of transformations. The tests will keep
it this way.

retronym commented:
@adriaanm I've addressed the comments, give it another spin when you have a moment.

One more question: do you know why `typer` widens method result types like this: ?

```
 % scala210 -nc -Xprint:typer -e 'def foo(a: String): a.type = a; val s: String = ""; foo(s): s.type' | grep foo
        private def foo(a: String): String = a;
         ($anon.this.foo($anon.this.s): this.s.type)
```

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1424/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1424/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2150/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2150/

adriaanm commented:
LGTM

indeed, packSymbols is the way to go

no idea regarding the widening of result types -- that seems wrong...

retronym commented:
> no idea regarding the widening of result types -- that seems wrong...

https://issues.scala-lang.org/browse/SI-6984 re-poses that question for our future selves.

adriaanm commented:
thanks

retronym commented:
Discussion with @odersky: he was supportive of the change, despite some concerns it involves adding casts, which is usually the responsibility of erasure. Maybe we would have been better deferring parameter list flattening until erasure. But we'd no doubt pay some tax in the intervening phases, specialization being among them.

It was also interesting to note that this problem only shows up in overriding checking (which leds to the missing bridge methods), which itself is a tricky, typeful business to pay the overloading tax.

Incidentally, that's the same class of problem we have associating extension methods with the original instance methods in the value class implementation, the subject of #1912.
